lib/trainer/test_parser.rb

```diff
@@ -120,6 +132,97 @@ def test_group_and_name(testable_summary, test, xcpretty_naming)
       return group, name
     end
 
+    def parse_xcresult(path)
+      # Executes xcresulttool to get JSON format of the result bundle object
+      result_bundle_object_raw = `xcrun xcresulttool get --format json --path #{path}`
+      result_bundle_object = JSON.parse(result_bundle_object_raw)
+
+      # Parses JSON into ActionsInvocationRecord to find a list of all ids for ActionTestPlanRunSummaries
+      actions_invocation_record = Trainer::XCResult::ActionsInvocationRecord.new(result_bundle_object)
+      test_refs = actions_invocation_record.actions.map do |action|
+        action.action_result.tests_ref
+      end.compact
+      ids = test_refs.map(&:id)
+
+      # Maps ids into ActionTestPlanRunSummaries by executing xcresulttool to get JSON
+      # containing specific information for each test summary,
+      summaries = ids.map do |id|
+        raw = `xcrun xcresulttool get --format json --path #{path} --id #{id}`
+        json = JSON.parse(raw)
+        Trainer::XCResult::ActionTestPlanRunSummaries.new(json)
+      end
+
+      # Converts the ActionTestPlanRunSummaries to data for junit generator
+      failures = actions_invocation_record.issues.test_failure_summaries
+      summaries_to_data(summaries, failures)
+    end
+
+    def summaries_to_data(summaries, failures)
+      # Gets flat list of all ActionTestableSummary
+      all_summaries = summaries.map(&:summaries).flatten
+      testable_summaries = all_summaries.map(&:testable_summaries).flatten
+
+      # Maps ActionTestableSummary to rows for junit generator
+      rows = testable_summaries.map do |testable_summary|
+        all_tests = testable_summary.all_tests.flatten
+
+        row = {
+          project_path: testable_summary.project_relative_path,
+          target_name: testable_summary.target_name,
+          test_name: testable_summary.name,
+
+          duration: all_tests.map(&:duration).inject(:+),
+          tests: all_tests.map do |test|
+            test_row = {
+              identifier: "#{test.parent.name}.#{test.name}",
+              name: test.name,
+              duration: test.duration,
+              status: test.test_status,
+              test_group: test.parent.name,
+
+              # These don't map to anything but keeping empty strings
+              guid: ""
+            }
+
+            # Tries to match failure on test case name
+            # Example:
+            # producing_target: "TestThisDude"
+            # test_case_name: "TestThisDude.testFailureJosh2()
+            found_failure = failures.find do |failure|
+              failure.test_case_name == test.identifier
+            end
+
+            # Set failure message if test status is "Failure" and a failure reference was found
+            if test.test_status == "Failure" && found_failure
```

Comment:
Is there a good place to put some of this logic in the xcresult models? If it's a bad fit, then here is fine. It just seems like theres a lot of data that the test_parser needs to know about.